trace/obfuscate.go

```diff
@@ -22,19 +55,148 @@ type obfuscationVisitor struct {
-	dp, isPrimitive := d.GetValue().(*pb.Data_Primitive)
-	if !isPrimitive {
+// EnterData processes the given data and obfuscates all the primitive values with zero values, regardless of it's meta data.
+func (*obfuscationVisitor) EnterData(self interface{}, _ vis.SpecVisitorContext, d *pb.Data) Cont {
+	dp := d.GetPrimitive()
+	if dp == nil {
 		return Continue
 	}
 
-	pv, err := spec_util.PrimitiveValueFromProto(dp.Primitive)
+	pv, err := spec_util.PrimitiveValueFromProto(dp)
 	if err != nil {
 		printer.Warningf("failed to obfuscate raw value, dropping\n")
 		d.Value = nil
 		return Continue
 	}
 
-	dp.Primitive.Value = pv.Obfuscate().ToProto().Value
+	dp.Value = pv.Obfuscate().ToProto().Value
+	return Continue
+}
+
+func (o *Obfuscator) RedactData(m *pb.Method) {
+	pov := redactSensitiveInfoVisitor{
+		obfuscationOptions: o,
+	}
+	vis.Apply(&pov, m)
+}
+
+type redactSensitiveInfoVisitor struct {
+	vis.DefaultSpecVisitorImpl
+	obfuscationOptions *Obfuscator
+}
+
+var _ vis.DefaultSpecVisitor = (*redactSensitiveInfoVisitor)(nil)
+
+// EnterData processes the given data and obfuscates sensitive information based on the provided obfuscation options.
+// It does 3 checks to determine if the data contains sensitive information:
+// 1. It checks if spec is an HTTP Authorization or Cookie data and obfuscates the value.
+// 2. It checks if spec is an HTTP Header or Query Param data and obfuscates the value if the key is in the list of sensitive keys.
+// 3. It applies regex patterns to all obfuscate sensitive primitive string values.
+func (s *redactSensitiveInfoVisitor) EnterData(self interface{}, _ vis.SpecVisitorContext, d *pb.Data) Cont {
+	if httpMeta := d.GetMeta().GetHttp(); httpMeta != nil {
+		var key string
+		switch httpMeta.Location.(type) {
+		case *pb.HTTPMeta_Auth:
+			return ObfuscatePrimitiveWithRedactedString(d)
+		case *pb.HTTPMeta_Cookie:
+			return ObfuscatePrimitiveWithRedactedString(d)
+		case *pb.HTTPMeta_Header:
+			header := httpMeta.GetHeader()
+			key = header.Key
+		case *pb.HTTPMeta_Query:
+			queryParam := httpMeta.GetQuery()
+			key = queryParam.Key
+		}
+		// Check if the key is in the list of keys to obfuscate.
+		if s.obfuscationOptions.SensitiveDataKeys.Contains(strings.ToLower(key)) {
+			return ObfuscatePrimitiveWithRedactedString(d)
+		}
+	}
+
+	primitive := d.GetPrimitive()
+	if primitive == nil {
+		// Not a primitive, regex will be applied to primitive string values only.
+		return Continue
+	}
+
+	stringValue := primitive.GetStringValue()
+	if stringValue == nil {
+		// Not a string, regex will be applied to string values only.
+		return Continue
+	}
+
+	for _, pattern := range s.obfuscationOptions.SensitiveDataValuePatterns {
+		if pattern.MatchString(stringValue.Value) {
+			return ObfuscatePrimitiveWithRedactedString(d)
+		}
+	}
+
+	return Continue
+}
+
+func (s *redactSensitiveInfoVisitor) EnterHTTPBody(self interface{}, ctx vis.SpecVisitorContext, b *pb.HTTPBody) Cont {
```

Comment:
As you mentioned, we need to handle HTTPMultipart, which should be treated the same as Body for redaction purposes.